clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitRepoArtifactCredentials.java

```diff
@@ -0,0 +1,150 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.artifacts.gitRepo;
+
+import com.netflix.spinnaker.clouddriver.artifacts.config.ArtifactCredentials;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.jgit.api.ArchiveCommand;
+import org.eclipse.jgit.api.CheckoutCommand;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.archive.ArchiveFormats;
+import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider;
+
+@Slf4j
+public class GitRepoArtifactCredentials implements ArtifactCredentials {
+  @Getter private final List<String> types = Collections.singletonList("git/repo");
+
+  @Getter private final String name;
+  private final String username;
+  private final String password;
+
+  public GitRepoArtifactCredentials(GitRepoArtifactAccount account) {
+    this.name = account.getName();
+    this.username = account.getUsername();
+    this.password = account.getPassword();
+  }
+
+  @Override
+  public InputStream download(Artifact artifact) throws IOException {
+    String repoReference = artifact.getReference();
+    Path stagingPath =
+        Paths.get(System.getProperty("java.io.tmpdir"), UUID.randomUUID().toString());
+
+    try {
+      log.info("Cloning git/repo {} into {}", repoReference, stagingPath.toString());
+      Git localRepository = clone(artifact, stagingPath);
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      log.info("Creating archive for git/repo {}", repoReference);
+      archiveToOutputStream(artifact, localRepository, outputStream);
+      return new ByteArrayInputStream(outputStream.toByteArray());
+    } catch (Exception e) {
+      throw new IOException("Failed to clone git/repo " + repoReference, e);
+    } finally {
+      log.info("Cleaning up git/repo {}", repoReference);
+      try {
+        FileUtils.deleteDirectory(stagingPath.toFile());
+      } catch (IOException e) {
+        log.info("Failed to cleanup git/repo for {}", repoReference);
+        throw new RuntimeException(
```

Comment:
In general it's not great to throw exceptions from within a `finally` block, primarily because it will mask any exception that happened in the `try` block...so if we fail to create the archive at all and throw an exception `Failed to clone git/repo...` then also fail to delete it there won't be any trace left of the original exception.

One workaround would be to use a try-with-resources construct:
```
try (AutoCloseable ignored = () -> FileUtils.deleteDirectory(stagingPath.toFile()))
```
and completely eliminate the `finally`.

Reply:
So, I think we'll still have the same problem using try-with-resources because `AutoCloseable.close` throws an exception which will need to be handled. In either case, to prevent the the main exception from being swallowed I think we'll just log and move on. I've taken a stab at implementing this. LMK what you think.

Reply:
One of the [advantages of try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) is that it will suppress the exception from `.close()` if (and only if) the `try` block threw an exception.  Always suppressing and logging the exception is also reasonable, but that will swallow the exception in cases where the `try` completes normally and we fail to close.